<a href="https://colab.research.google.com/github/laurahayas/Grupo_28_AYVD/blob/main/Entregable_Parte_2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>


**Diplomatura en Ciencia de Datos, Aprendizaje Automático y sus Aplicaciones**

**Edición 2022**

---

# Trabajo práctico entregable - Parte 2

In [ ]:
# Libraries
import io
import matplotlib
import matplotlib.pyplot as plt
import numpy
import pandas as pd
pd.set_option('display.float_format', '{:,.2f}'.format)
import seaborn as sns
sns.set_context('talk') 
import math
from scipy import stats



## Lectura del dataset
Vamos a utilizar la encuesta realizada en el anio 2020.

In [ ]:
url = 'https://cs.famaf.unc.edu.ar/~mteruel/datasets/diplodatos/sysarmy_survey_2020_processed.csv'
df = pd.read_csv(url)

In [ ]:
print(f'DataSet\n Filas: {df.shape[0]}\n Columnas: {df.shape[1]}')

DataSet
 Filas: 6095
 Columnas: 48


In [ ]:
df.head()

profile_gender  profile_age work_country                    work_province  \
0          Mujer           26    Argentina  Ciudad Autónoma de Buenos Aires   
1         Hombre           29    Argentina                       Corrientes   
2          Mujer           22    Argentina  Ciudad Autónoma de Buenos Aires   
3         Hombre           39    Argentina                       Corrientes   
4         Hombre           32    Argentina  Ciudad Autónoma de Buenos Aires   

   profile_years_experience  work_years_in_company  \
0                      3.00                   3.00   
1                      5.00                   2.00   
2                      2.00                   0.00   
3                     15.00                  15.00   
4                      9.00                   6.00   

   work_years_in_current_position  work_people_in_charge_of  \
0                            3.00                         0   
1                            2.00                         4   
2                            0.00                         0   
3                           13.00                         5   
4                            6.00                         0   

  profile_studies_level profile_studies_level_state  ...  \
0         Universitario                    En curso  ...   
1         Universitario                    En curso  ...   
2            Secundario                  Completado  ...   
3              Posgrado                  Incompleto  ...   
4         Universitario                    En curso  ...   

  work_has_violence_situations profile_sexual_orientation  \
0         En mi trabajo actual                 Homosexual   
1                        Jamás               Heterosexual   
2       En un trabajo anterior           Bisexual o queer   
3         En mi trabajo actual               Heterosexual   
4                        Jamás               Heterosexual   

  profile_has_disabilities profile_has_disabilities_hiring_difficulties  \
0                      NaN                                          NaN   
1                   Visual                                           No   
2                      NaN                                           No   
3                      NaN                                          NaN   
4                      NaN                                          NaN   

  company_employee_number                          company_main_activity  \
0                501-1000  Servicios / Consultoría de Software / Digital   
1                 201-500                               Otras industrias   
2               2001-5000                               Otras industrias   
3                501-1000                               Otras industrias   
4                   11-50  Servicios / Consultoría de Software / Digital   

  company_recommended company_diversity_policies  \
0                   7                          2   
1                   8                          9   
2                   6                          9   
3                   7                          6   
4                   6                          6   

                              company_extra_benefits  \
0  Capacitaciones y/o cursos, Comidas pagas / sub...   
1  Horarios flexibles, Stock options / RSUs, Viát...   
2  Clases de gimnasia online, Comidas pagas / sub...   
3  Abono de celular, Capacitaciones y/o cursos, C...   
4  Clases de idiomas, Horarios flexibles, Vacacio...   

  company_best_companies_city  
0                         NaN  
1                         NaN  
2                         NaN  
3                         NaN  
4                         NaN  

[5 rows x 48 columns]

In [ ]:
df.tail()

profile_gender  profile_age work_country  \
6090         Hombre           37    Argentina   
6091         Hombre           38    Argentina   
6092         Hombre           24    Argentina   
6093         Hombre           28    Argentina   
6094         Hombre           33    Argentina   

                        work_province  profile_years_experience  \
6090  Ciudad Autónoma de Buenos Aires                     13.00   
6091  Ciudad Autónoma de Buenos Aires                     12.00   
6092  Ciudad Autónoma de Buenos Aires                      2.00   
6093                         La Pampa                      5.00   
6094                              GBA                      5.00   

      work_years_in_company  work_years_in_current_position  \
6090                   6.00                            6.00   
6091                   5.00                            5.00   
6092                   2.00                            2.00   
6093                   3.00                            2.00   
6094                   5.00                            3.00   

      work_people_in_charge_of profile_studies_level  \
6090                         0             Terciario   
6091                         0         Universitario   
6092                         0             Terciario   
6093                         0             Terciario   
6094                         0             Terciario   

     profile_studies_level_state  ... work_has_violence_situations  \
6090                  Completado  ...         En mi trabajo actual   
6091                    En curso  ...                        Jamás   
6092                    En curso  ...                        Jamás   
6093                    En curso  ...                        Jamás   
6094                  Incompleto  ...                        Jamás   

     profile_sexual_orientation profile_has_disabilities  \
6090               Heterosexual                      NaN   
6091               Heterosexual                      NaN   
6092               Heterosexual                      NaN   
6093               Heterosexual                      NaN   
6094               Heterosexual                       no   

     profile_has_disabilities_hiring_difficulties company_employee_number  \
6090                                           No                501-1000   
6091                                          NaN               2001-5000   
6092                                          NaN                 201-500   
6093                                           No                  51-100   
6094                                           No                501-1000   

                              company_main_activity company_recommended  \
6090                               Otras industrias                   7   
6091                               Otras industrias                   8   
6092                    Producto basado en Software                   7   
6093  Servicios / Consultoría de Software / Digital                   5   
6094                               Otras industrias                   9   

     company_diversity_policies  \
6090                          8   
6091                         10   
6092                          9   
6093                          7   
6094                          7   

                                 company_extra_benefits  \
6090                                           Viáticos   
6091  Abono de celular, Capacitaciones y/o cursos, D...   
6092  Capacitaciones y/o cursos, Clases de idiomas, ...   
6093                          Capacitaciones y/o cursos   
6094   Abono de celular, Comidas pagas / subvencionadas   

     company_best_companies_city  
6090                         NaN  
6091                         NaN  
6092               Mercado Libre  
6093                 Solo Nubity  
6094                         NaN  

[5 rows x 48 columns]

In [ ]:
df[['profile_gender', 'salary_monthly_NETO']].groupby('profile_gender').describe()
# vemos que hay los valores minimos de la columna salarios son $1 (hombres), $ 2 (mujeres) y $ 40 (otros).
# vamos a filtrar el dataset para elimar datos que parecen estar mal ingresados.
#tb tenemos que sacar los sueldos mayores a 1000000

salary_monthly_NETO                                       \
                             count      mean        std   min       25%   
profile_gender                                                            
Hombre                    4,944.00 98,836.06 102,159.66  1.00 53,285.00   
Mujer                       908.00 73,501.41  56,256.88  2.00 43,818.00   
Otros                        31.00 95,441.35  88,635.67 40.00 39,500.00   

                                                  
                     50%        75%          max  
profile_gender                                    
Hombre         80,000.00 110,000.00 2,080,000.00  
Mujer          65,000.00  92,000.00   800,000.00  
Otros          74,000.00 117,500.00   380,000.00

Vamos aplicar un filtro a la columna 'salary_monthly_NETO' para eliminar los datos inferiores a $18.900 tal como explicamos en el Entregable 01 ese era el monto del salario minimo vital y movil para el periodo en que se realizo la encuesta por lo cual consideramos que podria ser un buen parametro para filtrar el dataset y eliminar del analisis las entradas asociadas a salarios extremadamente bajos y probablemente mal ingresados.
Ademas eliminamos los salarios superiores a 1000000


In [ ]:
# aplicamos el filtro al dataset
condition_1 = df.salary_monthly_NETO > 18900
condition_2 = df.salary_monthly_NETO <= 1000000

df_filtered = df[condition_1 & condition_2] 
df_filtered[['profile_gender', 'salary_monthly_NETO']].groupby('profile_gender').describe()

# confirmamos que luego de aplicar el filtro los salarios minimos son superiores al valor del salario_mvm

salary_monthly_NETO                                           \
                             count       mean       std       min       25%   
profile_gender                                                                
Hombre                    4,747.00 100,465.75 87,043.60 19,000.00 56,287.50   
Mujer                       848.00  78,487.33 54,865.09 20,000.00 49,000.00   
Otros                        28.00 105,660.71 87,235.71 34,000.00 43,750.00   

                                                  
                     50%        75%          max  
profile_gender                                    
Hombre         80,900.00 110,000.00 1,000,000.00  
Mujer          68,000.00  94,000.00   800,000.00  
Otros          78,900.00 122,750.00   380,000.00

In [ ]:
print(f'DataSet\n Filas: {df_filtered.shape[0]}\n Columnas: {df.shape[1]}')
# Vemos que se filtraron 464 entradas correspondientes con salarios menores a $18.900

DataSet
 Filas: 5620
 Columnas: 48


## Ejercicio 1: Estimación

**Consigna:**  Calcular una estimación puntual y un intervalo de confianza de nivel (1-alpha) para la resta entre la media del salario Neto para Hombres menos la media del salario Neto para otros géneros(diferencia de las medias entre el grupoA y grupoB). 
¿Cómo se relaciona este intervalo de confianza con el test de hipótesis?

In [ ]:
is_man = df_filtered.profile_gender == 'Hombre' 

groupA = df_filtered[is_man].salary_monthly_NETO
groupB = df_filtered[~is_man].salary_monthly_NETO

In [ ]:
print(type(groupA))
print(type(groupB))

<class 'pandas.core.series.Series'>
<class 'pandas.core.series.Series'>


In [ ]:
groupA_mean = groupA.mean()
groupA_std = groupA.std()
groupA_var = groupA.var()
print(f'Sueldo promedio Hombres = ${round(groupA_mean,2)}')
print(f'Std sueldo hombres = ${round(groupA_std,2)}')
print(f'Var sueldo hombres = ${round(groupA_var,2)}')

print('\n-----\n')

groupB_mean = groupB.mean()
groupB_std = groupB.std()
groupB_var = groupB.var()
print(f'Sueldo promedio Mujeres y otros = ${round(groupB_mean,2)}')
print(f'Std sueldo Mujeres y otros = ${round(groupB_std,2)}')
print(f'Var sueldo Mujeres y otros = ${round(groupB_var,2)}')


Sueldo promedio Hombres = $100465.75
Std sueldo hombres = $87043.6
Var sueldo hombres = $7576587857.24

-----

Sueldo promedio Mujeres y otros = $79355.88
Std sueldo Mujeres y otros = $56316.53
Var sueldo Mujeres y otros = $3171551196.51



### Estimacion puntual

In [ ]:
# diferencia entre salario promedio de hombres y otros generos (mujeres y otros)
dif_salario = groupA_mean - groupB_mean
print(f'Estimacion puntual de la diferencia entre salario de Hombres y Otros generos: \n${round(dif_salario,2)}')

Estimacion puntual de la diferencia entre salario de Hombres y Otros generos: 
$21109.86


### Intervalo de confianza
Si asumimos que las varianzas son iguales (algo que no pareciera, se podria hacer un test de homogeneidad de varianza) podemos utilizar el metodo de pivote. 

In [ ]:
# vamos a utilizar la std de toda la muestra y el n de toda la muestra
std_muestra = df_filtered.salary_monthly_NETO.std()
n_muestra = df_filtered.salary_monthly_NETO.count()
print(f'Std sueldo de toda la muestra = {round(std_muestra,2)}')
print(f'Tamanio de toda la muestra = {n_muestra}')

Std sueldo de toda la muestra = 80725.24
Tamanio de toda la muestra = 5620


In [ ]:
aplha = 0.05
z = stats.norm.ppf(1 - alpha/2)
print(f'z = {round(z,2)}')

z = 1.96


In [ ]:
IC_inferior = dif_salario - z * (std_muestra / math.sqrt(n_muestra))
IC_superior = dif_salario + z * (std_muestra / math.sqrt(n_muestra))
print(f'Intervalo de confianza >> [{round(IC_inferior,2)} , {round(IC_superior,2)}]')


Intervalo de confianza >> [18430.5 , 22651.54]


### Conslusion
Segun este resultado, asuminedo que las varianzas de los dos grupos muestrales son iguales,
la diferencia entre el sueldo promedio de hombres y otros esta dentro del intervalo [18430.5 , 22651.54] con un nivel de confianza del 95%.
Dado que el intervalo de confianza no incluye el valor 0 podemos inferir, con un nivel de confianza del 95% que el salario de los Hombres es superior al salario de otro generos.

---

### ¿Las varianzas son iguales?

In [ ]:
groupA_var - groupB_var
# a priori todo sugiere que las varianzas de los grupos no son iguales.
# vamos a realizar el test de Bartlett para confirmarlo

3897701441.532736

### Test de Bartlett
Para analizar si las varianzas de las poblaciones (Hombres y otros) son iguales o diferentes podemos utilizar el test de Bartlett

In [ ]:
# H0 = Los grupos tienen varianzas iguales
# H1 = Los grupos NO tienen varianzas iguales (o lo que es equivalente, los grupos tienen varianzas diferentes)
stats.bartlett(groupA, groupB)



BartlettResult(statistic=196.8428593092905, pvalue=1.0205104865339914e-44)

### Conclusion
Dado que pvalue=1.0205104865339914e-44 (p<0.05). Estamos en condiciones de rechazar H0. Por lo tanto los grupos NO tienen varianzas iguales.
Por lo tanto no seria correcto aplicar el metodo de pivote para calcular el intervalo de confianza para la diferencia entre la media del salario de hombres y otros generos.

---

## Ejercicio 2: Test de hipótesis



### 2.1 Formalización

Describir formalmente los distintos componentes de un test de hipótesis para comprobar si la distribución de los salarios es distinta entre los grupos A y B.

**Hipótesis**

H0 = La media del sueldo de los hombres es igual a la media del sueldo de mujeres y otros\
H1 = La media del sueldo de los hombres es distinta a la media del sueldo de mujeres y otros\
**Estadístico (Pivote)**
  * Identificar el estadístico >>> Promedio del sueldo
  * Escribir qué distribución tiene bajo $H_0$ >>> Normal (test de Shapiro, ver mas adelante)


### 2.2 P-valor

1. Calcule el p-valor y decida si rechazar o no la hipótesis nula.
2. Interprete el resultado.

Links útiles:
* [Test de hipótesis usando scipy](https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.ttest_ind.html)
* [Test de Welch](http://daniellakens.blogspot.com/2015/01/always-use-welchs-t-test-instead-of.html)

### Test de Shapiro
Vamos a utilizar el test de Shapiro para determinar si las muestras de sueldos de los dos grupos tiene o no distribucion normal

In [ ]:
# test de Shapiro para determinar si la muestra de sueldos de hombres tiene distribucion normal  
stats.shapiro(groupA)

# El resultado se lee asi (estadistico W, p-value).
# Como el p-value es menor a 0,05 podemos afirmar que la muestra tiene distribucion normal


(0.6183677315711975, 0.0)

In [ ]:
# test de Shapiro para determinar si la muestra de sueldos de otros tiene distribucion normal
stats.shapiro(groupB)

# El resultado se lee asi (estadistico W, p-value).
# Como el p-value es menor a 0,05 podemos afirmar que la muestra tiene distribucion normal

(0.6167652606964111, 3.4045386969850026e-40)

### Conclusion
En ambos casos el p-value es menor a 0,05. Por lo tanto podemos inferir, con nivel de confianza del 95% que las dos muestras tienen distribucion normal.

---

## Test de Welch

In [ ]:
# como los sueldos de ambos grupos tienen distribucion normal podemos aplicar el test de Welch.
# El test de Welch se aplica usando .ttest_ind(equal_var = False)
stats.ttest_ind(groupA, groupB, equal_var = False)

# como el p-value<0.05 podemos rechazar H0 (los sueldos de los dos grupos son iguales) y quedarnos con H1 (los sueldos de ambos grupos son diferentes)

Ttest_indResult(statistic=9.242572493015457, pvalue=6.749628469422985e-20)

### Conclusion
como el p-value<0.05 podemos rechazar H0 (los sueldos de los dos grupos son iguales) y quedarnos con H1 (los sueldos de ambos grupos son diferentes).

---

### [Opcional] 2.3 Potencia del test

Nuestra muestra, ¿era lo suficientemente grande para detectar si existe o no una diferencia entre los grupos?

1. Utilice la función `tt_ind_solve_power` para calcular el tamaño necesario de la muestra para un poder estadístico de 0.8, 0.9 y 0.95, asumiendo una significancia estadística de 0.05.
2. ¿Cómo intepretan el poder estadístico de un test? Dado su conocimiento de dominio sobre los datos, ¿les parece que esta muestra es lo suficientemente grande para ser representativo de la tendencia general? ¿y para utilizarlo en un juicio penal contra una empresa XX por una causa de discriminación?

[Documentación](https://www.statsmodels.org/stable/generated/statsmodels.stats.power.tt_ind_solve_power.html)

NOTA: este análisis debería hacerse ANTES de recolectar los datos.

## Ejercicio 3: Comunicación y visualización

**Consigna:** Seleccionen un resultado que les parezca relevante a partir de alguno de los ejercicios del entregable. Diseñe e implemente una comunicación en base a este mensaje, en un archivo PDF.

Elija las palabras y visualización más adecuada para que la comunicación sea entendible, efectiva y se dapte a UNA de las siguientes situaciones:

1. Una sección en un artículo de difusión a presentar como parte de una organización sin fines de lucro.
No más de 1 página A4 (o dos si los gráficos son muy grandes).
  1. Ejemplo: Alguna de las secciones [Los ecosistemas de emprendimiento de América Latina y el Caribe frente al COVID-19: Impactos, necesidades y recomendaciones](https://publications.iadb.org/es/los-ecosistemas-de-emprendimiento-de-america-latina-y-el-caribe-frente-al-covid-19-impactos-necesidades-y-recomendaciones), por ejemplo la sección *2.2. Reacciones de los emprendedores*.
  2. Ejemplo: Alguna de las secciones de [The state of gender pay gap in 2021](https://www.payscale.com/data/gender-pay-gap?tk=carousel-ps-rc-job)
  3. Puntos clave: 
    1. Simpleza de los gráficos.
    2. Comunicación en lenguaje simple a personas que no son necesariamente expertos de dominio.
    2. Selección de UNA oración sobre la que se hace énfasis.
    3. No es necesario que mencionen objetivos ni descripciones del conjunto de datos, se supone que eso ya estaría explicado en otras secciones del informe.

2. Una publicación científica o reporte técnico interno. No más de una página A4:
  2. Ejemplo: La sección de resultados de [IZA DP No. 12914: The Impact of a Minimum Wage Change on the Distribution of Wages and Household Income](https://www.iza.org/publications/dp/12914/the-impact-of-a-minimum-wage-change-on-the-distribution-of-wages-and-household-income).
  2. Ejemplo: Alguna de las secciones de [Temporary reduction in daily global CO2 emissions during the COVID-19 forced confinement](https://www.nature.com/articles/s41558-020-0797-x)
  3. Puntos clave:
    3. Nivel de detalle técnico requerido. Es necesario justificar la validez del análisis.
    4. La idea presentada puede ser más compleja. Pueden asumir que la audiencia tiene conocimiento técnico y va a analizar las visualizaciones en detalle. 
    5. Pueden presentar más en detalle las limitaciones del análisis (significancia estadística, etc.)
    2. No es necesario que mencionen objetivos ni descripciones del conjunto de datos, se supone que eso ya estaría explicado en otras secciones del informe.

3. Un tweet (o post de LinkedIn) para la cuenta de su empresa consultora que hace análisis de datos. El objetivo es promocionar un análisis de datos abiertos que van a incluir en su portfolio:
  1. Ejemplo: [Comparación vacunas covid](https://twitter.com/infobeautiful/status/1381577746527236098?s=20)
  2. Ejemplo: [Tweet del BID](https://twitter.com/el_BID/status/1388508583944507396?s=20). Lo valioso de este tweet es que usaron un único número para transmitir un mensaje. Puede ser algo así, o con un gráfico muy simple.
  3. Ejemplo: [Cambio climático](https://twitter.com/UNFCCC/status/1387732156190011394?s=20) Es un muy buen ejemplo, excepto que el gráfico no se lee nada y hay que entrar a la publicación original.
  3. Ejemplo: [¿Cuánto están los programadores en las empresas?](https://www.linkedin.com/posts/denis-rothman-0b034043_tech-career-work-activity-6793861923269054464-gS6y) (No verificamos la veracidad o seriedad de la fuente).
  4. Puntos clave:
    1. Su audiencia no va a mirar la visualización por más de unos segundos, y no tiene conocimiento técnico.
  3. Tienen que incluir además una *breve* descripción de cómo obtuvieron los datos que están presentando, que no entraría en el tweet.
